developers/weaviate/concepts/indexing/hnsw-indexes.mdx

```diff
@@ -0,0 +1,156 @@
+
+If the query limit is higher than `dynamicEfMax`, `dynamicEfMax` does not have any effect. In this case, the dynamic `ef` value is equal to the query limit.
+
+To determine the length of the ANN list, Weaviate multiples the query limit by `dynamicEfFactor`. The list range is modified by `dynamicEfMin` and `dynamicEfMax`.
+
+### Dynamic ef example
+
+Consider this GraphQL query that sets a limit of 4.
+
+```graphql
+{
+  Get {
+    JeopardyQuestion(limit: 4) {
+      answer
+      question
+    }
+  }
+}
+```
+
+Imagine the collection has dynamic `ef` configured.
+
+```json
+  "vectorIndexConfig": {
+     "ef": -1,
+     "dynamicEfMin": 5
+     "dynamicEfMax": 25
+     "dynamicEfFactor": 10
+  }
+```
+
+The resulting search list has these characteristics.
+
+- A potential length of 40 objects ( ("dynamicEfFactor": 10) * (limit: 4) ).
+- A minimum length of 5 objects ("dynamicEfMin": 5).
+- A maximum length of 25 objects ("dynamicEfMax": 25).
+- An actual size of 5 to 25 objects.
+
+#### Related environment variables
+
+If you use the [`docker-compose.yml` file from Weaviate](/developers/weaviate/installation/docker-compose) to run your local instance, the `QUERY_DEFAULTS_LIMIT` environment variable sets a reasonable default query limit.
+
+To prevent out of memory errors, the default setting for `QUERY_DEFAULTS_LIMIT` is significantly lower than `QUERY_MAXIMUM_RESULTS`.
+
+To change the default limit, edit the value for `QUERY_DEFAULTS_LIMIT` when you configure your Weaviate instance.
+
+## Deletions
+
+Deleted objects are marked as deleted, but Weaviate doesn't remove them immediately. Cleanup is an async process runs that rebuilds the HNSW graph after deletes and updates. During cleanup, Weaviate updates the graph and actually removes the deleted objects from the collection.
+
+Cleanup can affect system performance if your instance has a large number of deletes. These environment variables tune the cleanup process:
+
+- `TOMBSTONE_DELETION_CONCURRENCY`
+- `TOMBSTONE_DELETION_MAX_PER_CYCLE`
+- `TOMBSTONE_DELETION_MIN_PER_CYCLE`
+
+For more details, see [environment variables](/developers/weaviate/config-refs/env-vars).
+
+## Compression
+
+Consider using [compression](/developers/weaviate/starter-guides/managing-resources/compression) to reduce the size of for your HNSW indexes. Weaviate offers several ways to compress your data:
+
+import CompressionAlgorithms from '/_includes/starter-guides/compression-types.mdx';
+
+<CompressionAlgorithms/>
+
+## Other considerations
+
+HNSW indexes enable very fast queries, but they are not as fast at import time. Rebuilding the index when you add new vectors can be resource intensive. If you use HNSW, consider [enabling asynchronous indexing](/developers/weaviate/configuration/indexing-vector/dynamic-indexes#asynchronous-indexing) to improve system response during imports.
```

Comment:
> HNSW indexes enable very fast queries, but they are not as fast at import time.

Maybe we could say they incur "overhead to build the index" or similar. The first part of the sentence is absolute ("very fast queries") but the second is relative - but it's not super clear what it is being compared to. It almost reads as though imports are slower than queries, whereas I think we mean that building an HNSW index is slower than building a flat index.